From aaf012812e4e73ba8af5605d472329cdfb28c032 Mon Sep 17 00:00:00 2001
From: Stelios Tsampas <loathingkernel@gmail.com>
Date: Sat, 21 Mar 2026 18:50:34 +0200
Subject: [PATCH] upscalers: When checking for download links also check if the
 a cached file exists

Most likely the users already have the necessary dlls cached in the systems
from previous versions. In such a case, check if a dll with a matching name
already exists in the cache, and allow using a more recent version than the
one existing on the remove server.

This also has the side-effect of allowing users to provide their own
dlls in the cache to use with the downloader, instead of solely relying
on the remote to download them from. Users can now download their alternative
dlls and place them under `~/.cache/protonfixes/upscalers` with the following names

* amdxcffx64_v4.0.0_67A4D2BC10ad000.dll
* amdxcffx64_v4.0.1_67D435F7d97000.dll
* amdxcffx64_v4.0.2_68840348eb8000.dll
* amdxcffx64_v4.0.3_6930960536b9000.dll
* amdxcffx64_v4.1.0_69A0952A304a000.dll

each one representing their respective version. The can use the version
the want through `PROTON_FSR4_[_RDNA3]UPGRADE=<version>` where version one
of `4.0.0`, `4.0.1`, `4.0.2`, `4.0.3` or `4.1.0`.

This replacement was previously possible only for the `4.0.0` version.
---
 upscalers.py | 27 +++++++++++++++++++++++----
 1 file changed, 23 insertions(+), 4 deletions(-)

diff --git a/upscalers.py b/upscalers.py
index ee1326c..8d14d1d 100644
--- a/upscalers.py
+++ b/upscalers.py
@@ -153,14 +153,31 @@ def __get_fsr4_dlls(version: str = 'default') -> dict:
             'zip_md5_hash': None,
         },
     }
+
+    cache_dir = config.path.cache_dir.joinpath('upscalers')
+
+    def _cached_file_exists(file: dict) -> bool:
+        url_path = Path(unquote(urlparse(file['download_url']).path))
+        cached_file = cache_dir.joinpath(
+            url_path.stem + f'_v{file["version"]}' + url_path.suffix
+        )
+        return cached_file.exists()
+
     if version == 'default' or version not in __fsr4_dlls.keys():
-        version = '4.1.0'
-    if not __dll_download_exists(__fsr4_dlls[version]['download_url']):
+        version = '4.0.3'
+
+    item = __fsr4_dlls[version]
+    if not (__dll_download_exists(item['download_url']) or _cached_file_exists(item)):
         for key in sorted(__fsr4_dlls.keys(), reverse=True):
-            if __dll_download_exists(__fsr4_dlls[key]['download_url']):
+            item = __fsr4_dlls[key]
+            if __dll_download_exists(item['download_url']) or _cached_file_exists(item):
                 version = key
                 break
-    log.info(f'Using version {version} of amdxcffx64.dll')
+
+    log.info(
+        f'Found {"cached" if _cached_file_exists(item) else "remote"} version {version} of amdxcffx64.dll'
+    )
+
     return {
         'drive_c/windows/system32/amdxcffx64.dll': __fsr4_dlls[version],
     }
@@ -479,3 +496,5 @@ def setup_upscalers(
 
     if loaddll_replace:
         env['WINE_LOADDLL_REPLACE'] = ','.join(loaddll_replace)
+
+__all__ = ['setup_upscalers']
-- 
2.53.0

